pkg/actuators/machine/instances.go

```diff
@@ -201,22 +203,34 @@ func getBlockDeviceMappings(blockDeviceMappings []providerconfigv1.BlockDeviceMa
 	}
 	deviceName := describeAMIResult.Images[0].RootDeviceName
 
-	// Only support one blockDeviceMapping
-	volumeSize := blockDeviceMappings[0].EBS.VolumeSize
-	volumeType := blockDeviceMappings[0].EBS.VolumeType
-	blockDeviceMapping := ec2.BlockDeviceMapping{
-		DeviceName: deviceName,
-		Ebs: &ec2.EbsBlockDevice{
-			VolumeSize: volumeSize,
-			VolumeType: volumeType,
-			Encrypted:  blockDeviceMappings[0].EBS.Encrypted,
-		},
-	}
-	if *volumeType == "io1" {
-		blockDeviceMapping.Ebs.Iops = blockDeviceMappings[0].EBS.Iops
+	for _, blockDeviceMappingSpec := range blockDeviceMappingSpecs {
+		if blockDeviceMappingSpec.EBS == nil {
+			continue
+		}
+
+		if *blockDeviceMappingSpec.DeviceName != "" {
+			deviceName = blockDeviceMappingSpec.DeviceName
+		}
+
+		volumeSize := blockDeviceMappingSpec.EBS.VolumeSize
+		volumeType := blockDeviceMappingSpec.EBS.VolumeType
+
+		blockDeviceMapping := ec2.BlockDeviceMapping{
+			DeviceName: deviceName,
```

Comment:
This is going to attempt to give all devices the same name.

Reply:
I think it's expected that `DeviceName` should always be set apart from on the root device, though I don't think there's any validation on that at all since these aren't CRDs and we don't have schemas or webhooks for them.

Perhaps we can do validation here and return an error if more than one of the block device mappings doesn't have a name set?

Reply:
@JoelSpeed I thought of this but wasn't sure that's a right thing to do in that case

Reply:
Given we don't have any webhooks set up and we use a rawextension, I'm not sure what else we can do, I may well be missing something though, there may be a better way

Reply:
We don't need webhooks.  We can just fail the machine's reconcile and quit attempting.  I consider this an advanced usecase, so the user better follow the docs and set it up correctly.  This will probably also require investigation from the user to determine what a valid block device name is for their instance type.

Reply:
Coming back to this, we should make sure the PR that is adding webhooks should validate this, or we should follow up to update that (cc @enxebre)

Reply:
Captured this here -> https://github.com/openshift/machine-api-operator/pull/601/files#diff-98bc71b7347ada0763f9e57ae6c9d235R320
Regardless this should error and return and invalid config which the machine controller can use to fail. This seems to be case already